File: lib/fetch.js
const fs = require('fs');
const path = require('path');
const matter = require('gray-matter');
const { getAnnotations } = require('./annotate');

const CONTENT_DIR = path.join(__dirname, '..', 'content');

function readFile(dir, id) {
  const filepath = path.join(CONTENT_DIR, dir, `${id}.md`);
  if (!fs.existsSync(filepath)) {
    throw new Error(`File not found: ${dir}/${id}.md`);
  }
  return fs.readFileSync(filepath, 'utf8');
}

function listDir(dir) {
  const dirpath = path.join(CONTENT_DIR, dir);
  if (!fs.existsSync(dirpath)) return [];
  return fs.readdirSync(dirpath)
    .filter(f => f.endsWith('.md'))
    .map(f => f.replace('.md', ''));
}

function stripFrontmatter(content) {
  const parsed = matter(content);
  return parsed.content.trim();
}

// Canonical language names → the list of fenced-block tags that should be KEPT
// for that language. Lookup is case-insensitive after normalization.
// Aliases (py, js, ts) resolve to the same canonical as their full names.
const LANG_ALIASES = {
  py: 'python', python: 'python',
  js: 'javascript', javascript: 'javascript', node: 'javascript',
  ts: 'typescript', typescript: 'typescript',
  go: 'go', golang: 'go',
  rust: 'rust', rs: 'rust',
};
// What fenced-tag to treat as "the same language" for each canonical.
// TypeScript keeps both ts and js fences since the content currently only
// ships js examples (ts callers still benefit).
const LANG_FENCE_KEEPS = {
  python: ['python', 'py'],
  javascript: ['javascript', 'js', 'node'],
  typescript: ['typescript', 'ts', 'javascript', 'js'],
  go: ['go'],
  rust: ['rust', 'rs'],
};
// All fence tags recognized as "a code language we might filter".
const ALL_LANG_FENCES = new Set(['python', 'py', 'javascript', 'js', 'node', 'typescript', 'ts', 'go', 'rust', 'rs']);

function canonicalizeLang(lang) {
  if (!lang) return null;
  const normalized = lang.toLowerCase();
  const canonical = LANG_ALIASES[normalized];
  if (!canonical) {
    throw new Error(`Unknown language "${lang}". Known: py, python, js, javascript, ts, typescript, go, rust.`);
  }
  return canonical;
}

/**
 * Filter fenced code blocks by language. Keeps blocks whose fence tag is in
 * LANG_FENCE_KEEPS[canonical]. Removes blocks whose fence tag is in
 * ALL_LANG_FENCES but not in the keep list. Blocks with tags we don't
 * recognize (or no tag at all) are always preserved — prose remains untouched.
 *
 * Trailing blank lines left by a removed block are collapsed so the output
 * reads naturally when the agent consumes it.
 */
function filterLanguage(content, lang) {
  const canonical = canonicalizeLang(lang);
  if (!canonical) return content;
  const keepSet = new Set(LANG_FENCE_KEEPS[canonical]);

  // Match ```<tag>...\n<body>\n``` blocks. The tag is captured in group 1.
  const blockRe = /^```([A-Za-z0-9_+-]*)\s*\n[\s\S]*?^```\s*$\n?/gm;
  let out = content.replace(blockRe, (match, tag) => {
    const lowerTag = (tag || '').toLowerCase();
    // Unknown tags (empty, bash, shell, json, etc.) always preserved
    if (!ALL_LANG_FENCES.has(lowerTag)) return match;
    // Language-tagged block: keep or drop based on the language filter
    return keepSet.has(lowerTag) ? match : '';
  });

  // Collapse 3+ consecutive blank lines to 2 (from removed blocks)
  out = out.replace(/\n{3,}/g, '\n\n');
  return out;
}

// ─── Context filter ────────────────────────────────────────────────────
// Manifest-driven. The rules file lives at content/index/context-rules.json
// and declares, per-file, which patterns apply to which contexts. Patterns
// not listed in the manifest are always preserved.

const VALID_CONTEXTS = new Set(['web', 'agentic', 'rag', 'batch']);
let _contextRulesCache = null;

function loadContextRules() {
  if (_contextRulesCache) return _contextRulesCache;
  const rulesPath = path.join(CONTENT_DIR, 'index', 'context-rules.json');
  if (!fs.existsSync(rulesPath)) {
    _contextRulesCache = { rules: [] };
    return _contextRulesCache;
  }
  const raw = JSON.parse(fs.readFileSync(rulesPath, 'utf8'));
  _contextRulesCache = raw;
  return raw;
}

function canonicalizeContext(context) {
  if (!context) return null;
  const normalized = context.toLowerCase();
  if (!VALID_CONTEXTS.has(normalized)) {
    throw new Error(`Unknown context "${context}". Known: web, agentic, rag, batch.`);
  }
  return normalized;
}

/**
 * Given the full markdown content of an implement file, remove pattern
 * subsections whose heading is listed in the context rules with contexts
 * that do NOT include the requested context. A "pattern subsection" is a
 * `### Pattern N:` level-3 heading and everything until the next `### ` at
 * the same level (or the next `## ` section boundary, or EOF).
 *
 * Pattern headings with no rule in the manifest are preserved unconditionally.
 */
function filterContext(content, context, fileId) {
  const canonical = canonicalizeContext(context);
  if (!canonical) return content;

  const rules = loadContextRules().rules || [];
  const fileRules = rules.filter((r) => r.file === fileId);
  if (fileRules.length === 0) return content;

  const lines = content.split('\n');
  const keepLines = new Array(lines.length).fill(true);

  for (const rule of fileRules) {
    if (rule.contexts.includes(canonical)) continue; // pattern applies; keep
    // Find the heading line that starts with the prefix
    const headerRe = new RegExp('^###\\s+' + escapeRegex(rule.pattern_heading_prefix) + '\\b');
    for (let i = 0; i < lines.length; i++) {
      if (!headerRe.test(lines[i])) continue;
      // Mark lines from header until next ### or ## or EOF for removal
      let j = i;
      while (j < lines.length) {
        if (j > i && /^(##\s|###\s)/.test(lines[j])) break;
        keepLines[j] = false;
        j++;
      }
    }
  }

  const filtered = lines.filter((_, i) => keepLines[i]).join('\n');
  return filtered.replace(/\n{3,}/g, '\n\n');
}

function escapeRegex(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function appendAnnotations(content, fileId) {
  const annotations = getAnnotations(fileId);
  if (annotations.length === 0) return content;

  let annotationBlock = '\n\n---\n## Local annotations\n\n';
  for (const a of annotations) {
    annotationBlock += `- ${a.note} _(${a.date})_\n`;
  }
  return content + annotationBlock;
}

function get(fileIds, opts = {}) {
  if (opts.all) {
    fileIds = listDir('skills/areas');
  }

  if (!fileIds || fileIds.length === 0) {
    throw new Error('Provide file IDs or use --all. Run "aigis search --list" to see available files.');
  }

  const outputs = [];
  for (const id of fileIds) {
    let content = readFile('skills/areas', id);

    if (opts.frontmatter === false) {
      content = stripFrontmatter(content);
    }

    if (opts.lang) {
      content = filterLanguage(content, opts.lang);
    }

    if (opts.context) {
      content = filterContext(content, opts.context, id);
    }

    content = appendAnnotations(content, id);
    outputs.push(content);
  }

  return outputs.join('\n\n---\n\n');
}

function verify(fileIds) {
  if (!fileIds || fileIds.length === 0) {
    throw new Error('Provide file IDs to verify.');
  }

  const outputs = [];
  for (const id of fileIds) {
    const checklistId = `checklist-${id}`;
    const content = readFile('verify', checklistId);
    outputs.push(content);
  }

  return outputs.join('\n\n---\n\n');
}

function getTemplate(templateIds) {
  if (!templateIds || templateIds.length === 0) {
    throw new Error('Provide template IDs.');
  }

  const outputs = [];
  for (const id of templateIds) {
    const content = readFile('templates', id);
    outputs.push(content);
  }

  return outputs.join('\n\n---\n\n');
}

function getAuditScan() {
  return readFile('index', 'audit-scan');
}

function getCore() {
  return readFile('skills', 'core');
}

function getWorkflow(type) {
  if (!type) {
    throw new Error('Provide a workflow type. Run "aigis workflow --list" to see available workflows.');
  }
  return readFile('skills/workflows', type);
}

function listWorkflows() {
  return listDir('skills/workflows');
}

module.exports = { get, verify, getTemplate, getAuditScan, getCore, getWorkflow, listWorkflows, readFile, listDir };

File: cli-python/aigis_cli/fetch.py
from __future__ import annotations

import json
import re
from pathlib import Path

import frontmatter

from .annotate import get_annotations

CONTENT_DIR = Path(__file__).resolve().parent / "content"

SEPARATOR = "\n\n---\n\n"

# Language canonicalization — mirrors lib/fetch.js.
_LANG_ALIASES = {
    "py": "python", "python": "python",
    "js": "javascript", "javascript": "javascript", "node": "javascript",
    "ts": "typescript", "typescript": "typescript",
    "go": "go", "golang": "go",
    "rust": "rust", "rs": "rust",
}
_LANG_FENCE_KEEPS = {
    "python": {"python", "py"},
    "javascript": {"javascript", "js", "node"},
    "typescript": {"typescript", "ts", "javascript", "js"},
    "go": {"go"},
    "rust": {"rust", "rs"},
}
_ALL_LANG_FENCES = {"python", "py", "javascript", "js", "node", "typescript", "ts", "go", "rust", "rs"}

_VALID_CONTEXTS = {"web", "agentic", "rag", "batch"}
_context_rules_cache: dict | None = None


def read_file(directory: str, file_id: str) -> str:
    filepath = CONTENT_DIR / directory / f"{file_id}.md"
    if not filepath.exists():
        raise FileNotFoundError(f"File not found: {directory}/{file_id}.md")
    return filepath.read_text(encoding="utf-8")


def list_dir(directory: str) -> list[str]:
    dirpath = CONTENT_DIR / directory
    if not dirpath.exists():
        return []
    return sorted(
        p.stem for p in dirpath.iterdir() if p.suffix == ".md"
    )


def strip_frontmatter(content: str) -> str:
    parsed = frontmatter.loads(content)
    return parsed.content.strip()


def _canonicalize_lang(lang: str | None) -> str | None:
    if not lang:
        return None
    normalized = lang.lower()
    canonical = _LANG_ALIASES.get(normalized)
    if not canonical:
        raise ValueError(
            f'Unknown language "{lang}". Known: py, python, js, javascript, ts, typescript, go, rust.'
        )
    return canonical


def filter_language(content: str, lang: str | None) -> str:
    canonical = _canonicalize_lang(lang)
    if not canonical:
        return content
    keep_set = _LANG_FENCE_KEEPS[canonical]

    block_re = re.compile(r"^```([A-Za-z0-9_+-]*)\s*\n[\s\S]*?^```\s*$\n?", re.MULTILINE)

    def _sub(m: re.Match[str]) -> str:
        tag = (m.group(1) or "").lower()
        if tag not in _ALL_LANG_FENCES:
            return m.group(0)
        return m.group(0) if tag in keep_set else ""

    out = block_re.sub(_sub, content)
    return re.sub(r"\n{3,}", "\n\n", out)


def _load_context_rules() -> dict:
    global _context_rules_cache
    if _context_rules_cache is not None:
        return _context_rules_cache
    rules_path = CONTENT_DIR / "index" / "context-rules.json"
    if not rules_path.exists():
        _context_rules_cache = {"rules": []}
        return _context_rules_cache
    _context_rules_cache = json.loads(rules_path.read_text(encoding="utf-8"))
    return _context_rules_cache


def _canonicalize_context(context: str | None) -> str | None:
    if not context:
        return None
    normalized = context.lower()
    if normalized not in _VALID_CONTEXTS:
        raise ValueError(
            f'Unknown context "{context}". Known: web, agentic, rag, batch.'
        )
    return normalized


def filter_context(content: str, context: str | None, file_id: str) -> str:
    canonical = _canonicalize_context(context)
    if not canonical:
        return content

    rules = _load_context_rules().get("rules", [])
    file_rules = [r for r in rules if r.get("file") == file_id]
    if not file_rules:
        return content

    lines = content.split("\n")
    keep = [True] * len(lines)
    for rule in file_rules:
        if canonical in rule["contexts"]:
            continue  # pattern applies to this context; keep
        prefix = re.escape(rule["pattern_heading_prefix"])
        header_re = re.compile(r"^###\s+" + prefix + r"\b")
        i = 0
        while i < len(lines):
            if header_re.match(lines[i]):
                j = i
                while j < len(lines):
                    if j > i and re.match(r"^(##\s|###\s)", lines[j]):
                        break
                    keep[j] = False
                    j += 1
                i = j
                continue
            i += 1

    filtered = "\n".join(ln for ln, k in zip(lines, keep) if k)
    return re.sub(r"\n{3,}", "\n\n", filtered)


def append_annotations(content: str, file_id: str) -> str:
    annotations = get_annotations(file_id)
    if not annotations:
        return content
    block = "\n\n---\n## Local annotations\n\n"
    for a in annotations:
        block += f'- {a["note"]} _({a["date"]})_\n'
    return content + block


def get(file_ids: list[str] | None, *, all_files: bool = False,
        strip_fm: bool = True, lang: str | None = None,
        context: str | None = None) -> str:
    if all_files:
        file_ids = list_dir("implement")

    if not file_ids:
        raise ValueError('Provide file IDs or use --all. Run "aigis search --list" to see available files.')

    outputs: list[str] = []
    for fid in file_ids:
        content = read_file("implement", fid)
        if not strip_fm:
            content = strip_frontmatter(content)
        if lang:
            content = filter_language(content, lang)
        if context:
            content = filter_context(content, context, fid)
        content = append_annotations(content, fid)
        outputs.append(content)

    return SEPARATOR.join(outputs)


def verify(file_ids: list[str]) -> str:
    if not file_ids:
        raise ValueError("Provide file IDs to verify.")

    outputs: list[str] = []
    for fid in file_ids:
        content = read_file("verify", f"checklist-{fid}")
        outputs.append(content)

    return SEPARATOR.join(outputs)


def get_template(template_ids: list[str]) -> str:
    if not template_ids:
        raise ValueError("Provide template IDs.")

    outputs: list[str] = []
    for tid in template_ids:
        content = read_file("templates", tid)
        outputs.append(content)

    return SEPARATOR.join(outputs)


def get_audit_scan() -> str:
    return read_file("index", "audit-scan")
